### PR TITLE
Enable system smoke tests for all branches

### PR DESCRIPTION
**Pull request by dincho**
- reuse local docker image for system tests (cut ~8 minutes)
- install OTP from Erlang Solutions binary package (cut ~9 minutes)
- downgrade version to 20.1 to match minimum_otp_vsn in
rebar.config
- add machine build cache to CI (cut ~8 minutes)

As it can be seen in the CI workflow, system tests run a little bit faster than `test` step, thus is should not add to the total execution time in normal circumstances.

This introduces additional machine executor specific build cache, because of incompatibility with the docker executor cache. The cache is tried to be restored in the following order:
- commit hash: if the workflow is return without additional commits)
- branch: if new commits are pushed to the same branch, depending on the changes a portion of the cache is used
- master: a first push to a branch would restore the build cache from master, depending on the changes compared to the master a portion of the cache is used
- latest cache from any branch: until master cache is build at least once, e.g. once this PR is merged - it's cache will be used for the followed builds. Kind of warmup, can be removed later.

PT: https://www.pivotaltracker.com/story/show/160128182

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -93,6 +93,24 @@ references:
     restore_cache:
       key: *build_cache_key
 
+  machine_build_cache_key: &machine_build_cache_key machine-build-cache-v1-{{ .Branch }}-{{ .Revision }}
+  restore_machine_build_cache: &restore_machine_build_cache
```
Comment (lucafavatella) [inline .circleci/config.yml]: I see this cache not used for building Docker images ending in DockerHub. Ok.

File: .circleci/config.yml [hunk 2]
```diff
@@ -621,9 +645,8 @@ workflows:
               only: /^v.*$/
 
       - docker_system_smoke_tests:
-          filters:
```
Comment (lucafavatella) [inline .circleci/config.yml]:
```
          filters:
            branches:
              ignore:
                - env/dev1
                - env/dev2
```
?

File: .circleci/config.yml [hunk 3]
```diff
@@ -621,9 +645,8 @@ workflows:
...
-            branches:
-              only: master
+          requires:
+            - docker_smoke_tests
```
[the remark this answers is not in the document]
Comment (dincho) [reply .circleci/config.yml]: Yes, it's expected. I had trouble just tagging it, thus this solution is till ok.

File: .circleci/config.yml (hunk 2 above)
Comment (dincho) [reply .circleci/config.yml]: good point

Comment (dincho) [reply .circleci/config.yml]: add in eb114bb1

File: .circleci/config.yml [hunk 4]
```diff
@@ -93,6 +93,24 @@ references:
     restore_cache:
       key: *build_cache_key
 
+  machine_build_cache_key: &machine_build_cache_key machine-build-cache-v1-{{ .Branch }}-{{ .Revision }}
+  restore_machine_build_cache: &restore_machine_build_cache
+    restore_cache:
+      keys:
+        - *machine_build_cache_key
+        - machine-build-cache-v1-{{ .Branch }}-
+        - machine-build-cache-v1-master
```
[the remark this answers is not in the document]
Comment (dincho) [reply .circleci/config.yml]: changes in eb114bb1, however `master-foo` is still problematic. I have never seen such branch names.

[the remark this answers is not in the document]
Comment (lucafavatella) [reply .circleci/config.yml]: I may have used `master-check` or `master--check` in the past a couple of times (for re-triggering system tests on master).

**Pull request merged**